### PR TITLE
Normalize HTTP Lifecycle handlers with HTTP Probers

### PR DESCRIPTION
**Pull request by wzshiming**
#### What type of PR is this?
/kind feature
/kind bug
/sig node

#### Which issue(s) this PR fixes:

~xref kubernetes/enhancements#2771~

#### Does this PR introduce a user-facing change?

```release-note
Normalize HTTP lifecycle handlers with HTTP probers
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: pkg/kubelet/kubelet.go
Comment (matthyx) [inline pkg/kubelet/kubelet.go:496]: What's the purpose of enabling TLS and skip the certificate verifications?

**Review by matthyx — CHANGES_REQUESTED** [review]
Can you give more context?
- linked issue?
- why is it needed?
- what's the benefit?
- how to handle certificates properly?

Probably a KEP is required for something like that?

File: pkg/kubelet/kubelet.go
Comment (wzshiming) [reply pkg/kubelet/kubelet.go:496]: Originally, https was not supported. This is to support https and skip TLS, like Prober

**Comment by wzshiming** [comment]
I will write KEP later

**Comment by matthyx** [comment]
> I will write KEP later

Ok, this is just my opinion, maybe @ehashman or @SergeyKanzhelev think differently?

**Comment by wzshiming** [comment]
Hi @matthyx 
KEP is finished, could you have time to take a look?

https://github.com/kubernetes/enhancements/pull/2772

**Comment by aojea** [comment]
I think that we already discussed this in https://github.com/kubernetes/kubernetes/pull/99431#issuecomment-850284090
current probe already supports https
/hold

**Comment by wzshiming** [comment]
> I think that we already discussed this in [#99431 (comment)](https://github.com/kubernetes/kubernetes/pull/99431#issuecomment-850284090)
> current probe already supports https
> /hold

This is to fix Lifecycle httpGet does not support https, different from that

**Comment by wzshiming** [comment]
@aojea

Although the codes of `Lifecycle` and `Prober` are similar, they are indeed two different functions.

**Comment by aojea** [comment]
> > I think that we already discussed this in [#99431 (comment)](https://github.com/kubernetes/kubernetes/pull/99431#issuecomment-850284090)
> > current probe already supports https
> > /hold
> 
> @aojea
> This is to fix `Lifecycle` httpGet does not support https, different from that `Prober`
> Although the codes of `Lifecycle` and `Prober` are similar, they are indeed two different functions.

ups, sorry then, my bad 

But you still needs the KEP to get approved and tracked by the release team
https://github.com/kubernetes/enhancements#enhancements-tracking-spreadsheet

**Comment by wzshiming** [comment]
@aojea
Yes, I know that KEPs of 1.22 has been frozen, and the KEP Alpha release target v1.23
Wait for someone to see. could you have time to take a look? 😂

**Comment by aojea** [comment]
> > I will write KEP later
> 
> Ok, this is just my opinion, maybe @ehashman or @SergeyKanzhelev think differently?

hmm, @wzshiming I think that this is a bug, (if it is a bug there is no need for KEP)

The API already defines and scheme https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#httpgetaction-v1-core, defaults to HTTP but I can set it to HTTPS

If I create a pod setting HTTPS as the scheme of the lifecycle hook, it is allowed
```
spec:                                                                                                                                                                      
  containers:                                                                                                                                                              
  - image: nginx                                                                                                                                                           
    imagePullPolicy: Always                                                                                                                                                
    lifecycle:                                                                                                                                                             
      postStart:                                                                                                                                                           
        httpGet:                                                                                                                                                           
          path: healthz                                                                                                                                                    
          port: 8080                                                                                                                                                       
          scheme: HTTPS        
```

but the current requests are http instead of https, that IMHO is wrong, we allowed to set a scheme we should leverage it

Thiis is sig-node responsability and I may miss context in this area, I defer to @ehashman  @SergeyKanzhelev and @matthyx

**Comment by liggitt** [comment]
This looks like a duplicate of https://github.com/kubernetes/kubernetes/pull/86139 which has a lot of discussion about how to add support for this compatibly

**Comment by wzshiming** [comment]
> This looks like a duplicate of #86139 which has a lot of discussion about how to add support for this compatibly

hmm,  I thought it was a good catch, but I didn't expect it to be mentioned in 2019 ......... 😂😂😂

**Comment by wzshiming** [comment]
Hi @jasimmons
I picked your commit pick and added some additions, I think it’s time to fix this issue sooner.  WDYT @liggitt

**Comment by matthyx** [comment]
/lgtm
Bravo @wzshiming for sticking on it!

**Comment by liggitt** [comment]
This needs associated action required release notes, since it adds a feature gate and changes behavior of existing probes that are currently misconfigured but working. The "rollout / upgrade" considerations in a KEP are useful to think through and include so that existing cluster have a clear path to safely enabling this improvement

File: test/e2e/common/node/lifecycle_hook.go
Comment (liggitt) [inline test/e2e/common/node/lifecycle_hook.go:176]: 1.22?

File: test/e2e/common/node/lifecycle_hook.go
Comment (liggitt) [inline test/e2e/common/node/lifecycle_hook.go:180]: I don't think this can be in conformance yet. Existing clusters that have to transition non-disruptively will likely have to turn off the feature gate for at least one release, and with it off, will fail this e2e test (this is similar to the ExecTimeout support and e2e test in 1.20) 

also, this has a minimum kubelet version required, which needs to be reflected in the test tag

File: pkg/kubelet/lifecycle/handlers.go
Comment (liggitt) [inline pkg/kubelet/lifecycle/handlers.go:176]: for compatibility, should this fall back to http if the target does not support TLS?

Comment (aojea) [reply pkg/kubelet/lifecycle/handlers.go:176]: it seems to do it "automagically" 
```go
package main

import (
        "crypto/tls"
        "fmt"
        "io/ioutil"
        "log"
        "net/http"
)

func main() {

        http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
                fmt.Fprintf(w, "Hello")
        })
        log.Println("create server")

        go func() {
                log.Fatal(http.ListenAndServe("localhost:8081", nil))
        }()

        client := &http.Client{
                Transport: &http.Transport{
                        TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
                },
        }
        log.Println("request")
        req, err := http.NewRequest("GET", "http://localhost:8081", nil)
        if err != nil {
                panic(err)
        }

        response, err := client.Do(req)
        if err != nil {
                panic(err)
        }

        defer response.Body.Close()
        body, err := ioutil.ReadAll(response.Body)
        if err != nil {
                panic(err)
        }

        fmt.Println("Response", string(body))
}
```

Comment (wzshiming) [reply pkg/kubelet/lifecycle/handlers.go:176]: @aojea
``` go
req, err := http.NewRequest("GET", "https://localhost:8081", nil)
```
It should start with `https://` for TLS

Comment (aojea) [reply pkg/kubelet/lifecycle/handlers.go:176]: ah, great, mystery solved then, there is no auto fallback :sweat_smile:

**Comment by wzshiming** [comment]
@liggitt Updated, I added fallback and test. 


Also found that Lifecycle ignores StatusCodo. This is still different from Prober. After looking at the caller, I think it is necessary to add this.
https://github.com/kubernetes/kubernetes/blob/13cafd5cb00631e392fd4ce5d436097d6e885599/pkg/kubelet/kuberuntime/kuberuntime_container.go#L280-L290
https://github.com/kubernetes/kubernetes/blob/13cafd5cb00631e392fd4ce5d436097d6e885599/pkg/kubelet/kuberuntime/kuberuntime_container.go#L280-L290

**Comment by wzshiming** [comment]
Hi @liggitt could  have time to take a look

**Comment by aojea** [comment]
> Also found that Lifecycle ignores Response's StatusCode. This is still different from Prober. After looking at the caller, I think it is necessary to add this.

does this modifies previous behaviour?

**Comment by wzshiming** [comment]
> > Also found that Lifecycle ignores Response's StatusCode. This is still different from Prober. After looking at the caller, I think it is necessary to add this.
> 
> does this modifies previous behaviour?

Yes, There is a change in expected behavior. The Response's StatusCode >=400 is unprocessed before, and it is regarded as successful by default.

https://github.com/kubernetes/kubernetes/pull/102474/commits/8e7863f82a5ece8e0517b69dc19001a7d5d47207

Follow Prober:
https://github.com/kubernetes/kubernetes/blob/59f150333ed0edc092587c275d16455aafa5d441/pkg/probe/http/http.go#L129-L138

**Comment by liggitt** [comment]
> Yes, There is a change in expected behavior. The Response's StatusCode >=400 is unprocessed, and it is regarded as successful by default.
> 
> Maybe add FeatureGate separately for the changed expected behavior

Any change that could make an existing successful lifecycle handler unsuccessful after upgrade is extremely difficult to roll out and should not be done lightly. What we've done in other similar scenarios:

1. identify the behavior change that is potentially breaking
2. make those behavior changes optional and able to be reverted, likely with a feature gate
3. instrument the area with metrics so that if a configuration or execution would have been affected by the breaking change, it can be detected via metrics
4. wait until we have positive confirmation from a broad swath of deployments that the metrics and optional change is acceptable before making the change permanent

**Comment by wzshiming** [comment]
OK, then I remove this part first and open a new issue to discuss this?

**Pull request closed without merging**
